### ISSUE TITLE
[2018 CVPR] Non-local Neural Networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#44]
Non-local block은 self-attention을 추상화한 식이며 spatial,  temporal attention로 사용 가능

논문 내용 중 non-local, experiment만 설명

### 1.1. Formulation
<img src='https://user-images.githubusercontent.com/49643709/147706753-d4a4e5ea-75d6-4522-aba1-244ae60807db.png' width=80% height=80%>

- i : index of an output position
- j : index that enumerates all possible positions
- x : input signal
- y : output signal of the same size as x
- f(x_i, x_j) : pairwise function that computes a scalar between i and all j (representing relationship such as affinity)
- g(x_j) : unary function that computes a representation input signal at the position j
- C(x) : normalize factor

### 1.2. Instantiations
<img src='https://user-images.githubusercontent.com/49643709/147707299-7cc3f147-b665-4886-94ba-06b84cbb0aa3.png' width=80% height=80%>

간편성을 위해 unary function g를 linear embedding으로 고정

Pairwise function f를 다양하게 세팅하여 실험 진행

#### 1.2.1. Gaussian
<img src='https://user-images.githubusercontent.com/49643709/147707518-2cd60cc3-6c7b-4203-805f-54738379d3b0.png' width=80% height=80%>

#### 1.2.2. Embedded Gaussian
<img src='https://user-images.githubusercontent.com/49643709/147707583-2c6d3a97-ea32-4313-9cb3-cbed617ae955.png' width=80% height=80%>

- Self-attention module과 동일

#### 1.2.3. Dot product
<img src='https://user-images.githubusercontent.com/49643709/147707629-36a3bec4-c2ef-4d3d-b89d-782449cc17c0.png' width=80% height=80%>

- Embedded Gaussian version과의 차이점은 softmax 존재 여부 (softmax를 사용하면 f의 output은 양수)

#### 1.2.4. Concatenation
<img src='https://user-images.githubusercontent.com/49643709/147707727-4e2803a1-5eb7-4a13-a765-0bef385de4dd.png' width=80% height=80%>

### 1.3. Non-local Block
<img src='https://user-images.githubusercontent.com/49643709/147707830-f2bcb8e2-8131-4eaa-b5c0-f7ee9acc0ad5.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/147707862-463e41b1-541d-475f-8396-2294d5e69d23.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/147707964-667cfada-c696-4876-9421-c0d7ca48e39d.png' width=80% height=80%>

- ResNet의 residual과 같은 방식으로 non-local block 사용
- Subsampling trick을 이용하여 computation 줄임 (2*2 max pooling)

## 2. Experiments

대부분의 ablation study는 video에서 실험했기에, image에서 다소 안 맞을 수 있음

### 2.1. Video
![image](https://user-images.githubusercontent.com/49643709/147771245-7704e05c-b35a-4319-a74f-82c83a42103d.png)

### 2.1.1. Instantiations - Table 2(a)
- 기존 모델에 res4 last block 전에 1개의 non-local block 추가
- Pairwise function을 다양하게 써도 성능 개선이 있었으며, 그 수치가 유사함
- 즉 attentional behavior가 중요한 것이 아니라 non-local behavior가 중요하다는 것을 보여줌

### 2.1.2. Which stage to add non-local blocks - Table 2(b)
- Res2, 3, 4에 non-local block을 추가하면 성능이 비슷하게 오름
- Res5의 경우 성능이 다소 떨어지는데 spatial size가 작아 precise spatial information을 제공하기에 어렵기 때문

### 2.1.3. Going deeper with non-local blocks - Table 2(c)
- Non-local block을 추가할 수록 성능이 더 좋아짐
- Layer depth가 추가되어 성능이 증가하는 것이 아니라 non-local 때문에 증가함 (R50 with 5 non-local > R101)

#### 2.1.4. Performance
![image](https://user-images.githubusercontent.com/49643709/147771601-aa8281e1-ad9d-4774-b05e-2e39a7ec8557.png)
<img src='https://user-images.githubusercontent.com/49643709/147771326-6b0cfab9-ccab-4e2e-b3a9-3c7604f31681.png' width=60% height=60%>

- 기존에 비해 좋음

#### 2.2.1. Object detection and instance segmentation
<img src='https://user-images.githubusercontent.com/49643709/147778405-f6ad7784-4390-4b90-9b47-9f3d2d03d9ce.png' width=80% height=80%>

- Mask R-CNN backbone last residual block of res4 전에 single non-local block 추가했더니 성능 증가
- 즉 기존의 모델들이 increased depth/capacity로 성능을 올려왔지만, non-local dependency를 충분히 못잡았다는 것임
- Backbone에 더 많이 non-local block을 추가하면 오히려 감소하는 현상이 보임 (video와 달리)

#### 2.2.2. Keypoint detection
<img src='https://user-images.githubusercontent.com/49643709/147787917-c04188af-81ab-4071-b143-ebb2ae49cfe0.png' width=80% height=80%>

**Comment by Jasonlee1995** [comment]
[nonlocal.pdf](https://github.com/Jasonlee1995/AI_Papers/files/7798630/nonlocal.pdf)
